File: packages/node/tests/unit/issuer.test.ts
import { describe, expect, it } from "vitest";
import {
  assertTenantAllowed,
  isTenantGuid,
  isValidTenantId,
  PERSONAL_MSA_TENANT_ID,
  resolveExpectedIssuer,
} from "../../src/providers/microsoft/issuer.js";
import { validateMicrosoftConfig } from "../../src/config/validate.js";
import { InvalidIssuerError } from "../../src/errors/index.js";
import { CLIENT_ID, TENANT_ID, OTHER_TENANT_ID } from "../helpers/crypto.js";

const baseConfig = (tenantId: string, allowedTenants?: string[]) =>
  validateMicrosoftConfig({
    clientId: CLIENT_ID,
    tenantId,
    ...(allowedTenants ? { allowedTenants } : {}),
    fetch: () => Promise.reject(new Error("unused")),
  });

describe("tenant id helpers", () => {
  it("recognizes GUIDs", () => {
    expect(isTenantGuid(TENANT_ID)).toBe(true);
    expect(isTenantGuid("common")).toBe(false);
  });

  it("validates named modes and GUIDs", () => {
    for (const mode of ["common", "organizations", "consumers", TENANT_ID]) {
      expect(isValidTenantId(mode)).toBe(true);
    }
    expect(isValidTenantId("nonsense")).toBe(false);
  });
});

describe("resolveExpectedIssuer", () => {
  it("builds the v2 issuer from the real tenant id", () => {
    expect(resolveExpectedIssuer(TENANT_ID, "2.0", baseConfig("common"))).toBe(
      `https://login.microsoftonline.com/${TENANT_ID}/v2.0`,
    );
  });

  it("builds the v1 issuer with trailing slash", () => {
    expect(resolveExpectedIssuer(TENANT_ID, "1.0", baseConfig("common"))).toBe(
      `https://sts.windows.net/${TENANT_ID}/`,
    );
  });
});

describe("assertTenantAllowed", () => {
  it("allows any tenant under 'common'", () => {
    expect(() => assertTenantAllowed(OTHER_TENANT_ID, baseConfig("common"))).not.toThrow();
  });

  it("requires an exact match under a GUID tenant", () => {
    expect(() => assertTenantAllowed(TENANT_ID, baseConfig(TENANT_ID))).not.toThrow();
    expect(() => assertTenantAllowed(OTHER_TENANT_ID, baseConfig(TENANT_ID))).toThrow(
      InvalidIssuerError,
    );
  });

  it("rejects personal accounts under 'organizations'", () => {
    expect(() => assertTenantAllowed(TENANT_ID, baseConfig("organizations"))).not.toThrow();
    expect(() => assertTenantAllowed(PERSONAL_MSA_TENANT_ID, baseConfig("organizations"))).toThrow(
      InvalidIssuerError,
    );
  });

  it("requires the MSA tenant under 'consumers'", () => {
    expect(() =>
      assertTenantAllowed(PERSONAL_MSA_TENANT_ID, baseConfig("consumers")),
    ).not.toThrow();
    expect(() => assertTenantAllowed(TENANT_ID, baseConfig("consumers"))).toThrow(
      InvalidIssuerError,
    );
  });

  it("enforces an explicit allow-list on top of the mode", () => {
    const config = baseConfig("common", [TENANT_ID]);
    expect(() => assertTenantAllowed(TENANT_ID, config)).not.toThrow();
    expect(() => assertTenantAllowed(OTHER_TENANT_ID, config)).toThrow(InvalidIssuerError);
  });
});

File: packages/go/provider/microsoft/issuer_test.go
package microsoft

import (
	"testing"

	"github.com/avi-pathak/easy-sso/packages/go/config"
	"github.com/avi-pathak/easy-sso/packages/go/ssoerr"
)

func norm(t *testing.T, cfg config.MicrosoftAuthConfig) config.NormalizedMicrosoftConfig {
	t.Helper()
	n, err := config.ValidateMicrosoftConfig(cfg)
	if err != nil {
		t.Fatalf("validate: %v", err)
	}
	return n
}

func TestResolveExpectedIssuer(t *testing.T) {
	cfg := norm(t, config.MicrosoftAuthConfig{ClientID: "c", TenantID: "common"})
	tid := "abc"
	if got := ResolveExpectedIssuer(tid, config.TokenV2, cfg); got != "https://login.microsoftonline.com/abc/v2.0" {
		t.Fatalf("v2 issuer = %q", got)
	}
	if got := ResolveExpectedIssuer(tid, config.TokenV1, cfg); got != "https://sts.windows.net/abc/" {
		t.Fatalf("v1 issuer = %q", got)
	}
}

func TestAssertTenantAllowed(t *testing.T) {
	guid := "11111111-2222-3333-4444-555555555555"
	other := "99999999-2222-3333-4444-555555555555"

	cases := []struct {
		name    string
		cfg     config.MicrosoftAuthConfig
		tid     string
		wantErr bool
	}{
		{"guid match", config.MicrosoftAuthConfig{ClientID: "c", TenantID: guid}, guid, false},
		{"guid mismatch", config.MicrosoftAuthConfig{ClientID: "c", TenantID: guid}, other, true},
		{"common allows any", config.MicrosoftAuthConfig{ClientID: "c", TenantID: "common"}, other, false},
		{"organizations rejects personal", config.MicrosoftAuthConfig{ClientID: "c", TenantID: "organizations"}, config.PersonalMSATenantID, true},
		{"organizations allows work", config.MicrosoftAuthConfig{ClientID: "c", TenantID: "organizations"}, guid, false},
		{"consumers requires personal", config.MicrosoftAuthConfig{ClientID: "c", TenantID: "consumers"}, guid, true},
		{"allow-list miss", config.MicrosoftAuthConfig{ClientID: "c", TenantID: "common", AllowedTenants: []string{guid}}, other, true},
		{"allow-list hit", config.MicrosoftAuthConfig{ClientID: "c", TenantID: "common", AllowedTenants: []string{guid}}, guid, false},
	}
	for _, tc := range cases {
		t.Run(tc.name, func(t *testing.T) {
			err := AssertTenantAllowed(tc.tid, norm(t, tc.cfg))
			if tc.wantErr {
				if err == nil {
					t.Fatal("expected error")
				}
				if ae, ok := ssoerr.As(err); !ok || ae.Code != ssoerr.CodeInvalidIssuer {
					t.Fatalf("want invalid_issuer, got %v", err)
				}
			} else if err != nil {
				t.Fatalf("unexpected error: %v", err)
			}
		})
	}
}
